### ISSUE TITLE
can you tell me why you do that ? (about //)

### ISSUE THREAD

**Issue by rexhxiao** [Neilski/URF-Identity#1]
public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken)
        {
            // SyncObjectsStatePreCommit();
            var changesAsync = await base.SaveChangesAsync(cancellationToken);
            // SyncObjectsStatePostCommit();
            return changesAsync;
        }

**Comment by Neilski** [comment]
See https://genericunitofworkandrepositories.codeplex.com/discussions/645729 - is this what you mean?

**Comment by rexhxiao** [comment]
sorry,my english is very poor 
i am install urf from nuget and asp.net identity go together
then i find you here

in my project,the "CreateNewUser" doesn't work for original urf assembly in AccountController

    var result = await UserManager.CreateAsync(user, model.Password);

result alway success, but database have no data
I tested because user ObjectState,I mean, just do

    user.ObjectState = ObjectState.Added;

To Be able to works,i am find difference between us from your project
the difference is you using comments as follows

    public override async Task<int> SaveChangesAsync(CancellationToken cancellationToken)
    {
        // SyncObjectsStatePreCommit();
        var changesAsync = await base.SaveChangesAsync(cancellationToken);
        // SyncObjectsStatePostCommit();
        return changesAsync;
    }

i don't know why do this 
and,if it is possible use by original urf assembly?

**Comment by Neilski** [comment]
Hello, no problem.

To answer your question, it seems like there has been a change to the library since I published my project.  I will update to the latest build, test and come back to you.

Neil.

**Comment by rexhxiao** [comment]
Thx a lot. 

I look forward to your reply, i was like a cat on a hot tin roof.

**Comment by Neilski** [comment]
It turns out that this was quite a major bug in the implementation.  In summary, the Identity 2.0 management and update methods do not (of course!) manage the URF's IObjectState correctly, so when changes were saved (DataContext.SaveChanges() or DataContext.SaveChangesAsync()) the change in entity were not detected - because IObjectState had not been set.

I am not totally happy with the solution I have implemented because it does mean modifying the URF library (specifically the Repository.Pattern.Ef6.DataContext.cs) and it also means that you have to manage Identity derived entities within the application's DataContext implementation as a 'special case' but, for now at least, I coudn't think of a better solution.

I have updated the project with the code changes.  Let me know if this works for you or if you have a  better, more elegant solution.

**Comment by rexhxiao** [comment]
I am appreciated greatly your help.

In my project, I use a temporary solution to this problem, because it only affects the AspNet.Identity and I do not want to modify the URF library.

    public class ApplicationUserManager : UserManager<ApplicationUser, int>
    {
        public ApplicationUserManager(IUserStore<ApplicationUser, int> store)
            : base(store)
        {
        }

        public override Task<IdentityResult> CreateAsync(ApplicationUser user)
        {
            user.ObjectState = ObjectState.Added;
            return base.CreateAsync(user);
        }

        public override Task<IdentityResult> UpdateAsync(ApplicationUser user)
        {
            user.ObjectState = ObjectState.Modified;
            return base.UpdateAsync(user);
        }
    }
The above is provided for your reference

**Comment by Neilski** [comment]
I started off with this approach too, but there are other areas of the Identity Framework that update the ApplicationUser and ApplicationRole entities.  I didn't like modifying the URF library either, but in this case, it seemed like the best option.

I've been playing around and I actually think my posted solution can be simplified further, buy just modifying the SyncObjectsStatePreCommit() as follows:

```
private void SyncObjectsStatePreCommit()
{
    // The body of the following foreach() loop originally simply assigned the
    // urfEntityState to the dbEntityEntry.State.  Unfortunately, this approach
    // failed with Identity 2.0 managed entities.
    //
    // Microsoft's Identity 2.0 framework knows nothing about the URF's
    // IObject state and works directly with the DnEntityEntry state so the URF
    // is unaware of any changes made to entities managed by Identity 2.0.
    foreach (var dbEntityEntry in ChangeTracker.Entries())
    {
        var urfEntityState =
            StateHelper.ConvertState(
                ((IObjectState) dbEntityEntry.Entity).ObjectState);

        if ((dbEntityEntry.State == EntityState.Unchanged) &&
            (urfEntityState != dbEntityEntry.State))
        {
            dbEntityEntry.State = urfEntityState;
        }
    }
}
```

I then renamed the DataContext.cs file to DataContext-MODIFIED.cs to try and make sure I don't overlook it if I update the URF in the future.  You could also add a compiler pragma to provide you with a constant warning that you have modified library code.

This needs some proper testing, but it seems to work and, if it does, it would also support any non-URF aware frameworks or services that you might plug in to your application.

**Comment by Neilski** [comment]
I guess that if you did the create, update and delete methods in both the overridden ApplicationUser and ApplicationRole classes you should probably be OK - this might even be a better route to go.  How much testing have you done?

**Comment by rexhxiao** [comment]
Yeah , you're right ! there are other areas of the Identity Framework that update the ApplicationUser and ApplicationRole entities.

I would be glad to do it .If I have found any problem, I will inform you immediately.

I just started not long after, there are not many test.

**Comment by Neilski** [comment]
Thank you.  I am not sure whether it is better to override the ApplicationUserManager/ApplicationRoleManager or the ApplicationUserStore/ApplicationRoleStore classes, or both.  I am running some basic test now.

**Comment by Neilski** [comment]
It seems that just implementing this in the ApplicationUserStore and ApplicationRoleStore appears to be sufficient...

```
public class ApplicationUserStore : UserStore<ApplicationUser, ApplicationRole, Guid,
    ApplicationUserLogin, ApplicationUserRole, ApplicationUserClaim>
{
    public ApplicationUserStore(IUrfIdentityDataContextAsync context)
        : base(context as DbContext)
    {
    }

    public override Task CreateAsync(ApplicationUser user)
    {
        user.ObjectState = ObjectState.Added;
        return base.CreateAsync(user);
    }

    public override Task UpdateAsync(ApplicationUser user)
    {
        user.ObjectState = ObjectState.Modified;
        return base.UpdateAsync(user);
    }

    public override Task DeleteAsync(ApplicationUser user)
    {
        user.ObjectState = ObjectState.Deleted;
        return base.DeleteAsync(user);
    }
}
```
and
```
public class ApplicationRoleStore :
    RoleStore<ApplicationRole, Guid, ApplicationUserRole>
{
    public ApplicationRoleStore(IUrfIdentityDataContextAsync context)
        : base(context as DbContext)
    {
    }


    public override Task CreateAsync(ApplicationRole role)
    {
        role.ObjectState = ObjectState.Added;
        return base.CreateAsync(role);
    }


    public override Task UpdateAsync(ApplicationRole role)
    {
        role.ObjectState = ObjectState.Modified;
        return base.UpdateAsync(role);
    }


    public override Task DeleteAsync(ApplicationRole role)
    {
        role.ObjectState = ObjectState.Deleted;
        return base.DeleteAsync(role);
    }
}
'''

Seems to be sufficient.

**Comment by Neilski** [comment]
I have updated the project based upon your suggestions - thank you for your help - this feels like a much better solution.

I'll leave this ticket open for a while until we are confident that the solution is stable.

**Comment by rexhxiao** [comment]
It’s my pleasure, let us together diligently to better.

**Comment by saYRam** [comment]
var addLogin = await UserManager.AddLoginAsync(user.Id, info.Login) 
and
UserManager.UpdateAsync(user);

also not woking. Does not store the data to database. Is it fixed?

**Comment by Neilski** [comment]
Have you downloaded and referenced the latest build?

**Comment by Neilski** [comment]
I have just tested the latest build as follows:

1. Download the latest build (Download ZIP) and decompressed
2. Opened the solution in Visual Studio 2015
3. Modified the web.config file to define the SMTP service for user verification testing
4. Checked the .mdf database file in /App-Data to make sure the AspNetUsers table was empty
5. Completed a user registration and responded to the verification email
6. re-checked the .mdf file - the AspNetUsers table now contains the details for my newly registered user.

Please complete the above steps and let me know if you still have a problem.

**Comment by saYRam** [comment]
No no. I'm not talking about normal registration. It's works fine. But there is problem on external logins. I downloaded project again (15 mins ago) and tested Google external auth. When the user authorized by google, registration complated but external login details (providerkey etc) not stored to the database. There is no error because await UserManager.AddLoginAsync(user.Id, info.Login) returns succeeced message.

        public async Task<ActionResult> LinkLoginCallback()
        {
            var loginInfo =
                await
                    AuthenticationManager.GetExternalLoginInfoAsync(
                        XsrfKey, User.Identity.GetUserId());

            if (loginInfo == null)
            {
                return RedirectToAction(
                    "ManageLogins", new {Message = ManageMessageId.Error});
            }

            var userId = Guid.Parse(User.Identity.GetUserId());

            var result =
                await
                    UserManager.AddLoginAsync(userId, loginInfo.Login);

            return result.Succeeded
                ? RedirectToAction("ManageLogins")
                : RedirectToAction(
                    "ManageLogins", new {Message = ManageMessageId.Error});
        }

in runtime with this code everything seems works fine. result return Secceeced = true. But AspNetUserLogins table still empty.

In my another project also cant add user to role. UserManager.AddToRoleAsync is return succeced but its not. Even UserManager.UpdateAsync does not works. Never throws error, at run time everything seems fine but no effect.

**Comment by Neilski** [comment]
I have not tested that scenario, but I suspect that you will need to override another method in the Identity.ApplicationUserStore in the project.  I will look at this when I get back to the office and hopefully find a fix.

**Comment by Neilski** [comment]
Update.  I believe I have fixed the problem with the external login and will update the project tomorrow.  The problem with the UserManager.AddToRoleAsync() method is proving more difficult as it appears the the new assignment record is being created in the UserStore itself (rather than in the UserManager code).

**Comment by Neilski** [comment]
After digging into the Identity source code, I have realised that simply coercing the ObjectState in the various Identity data store overrides does not completely address the problem of syncing EF's EntityState to URF's ObjectState.

In most scenarios it works as expected, but for User sub-properties such as role assignment, this processing is actually implemented within the Identity UserStore itself, so there appears to be no opportunity to set the ObjectState - although I am happy to be proven wrong on this!

To get over this problem, I have resorted to something I had hoped to avoid in ***modifying the core URF library code*** - specifically modifying the SyncObjectsStatePreCommit() method in the Repository.Pattern.Ef6.DataContext.cs file.

Please try downloading the code and testing again.  Please note that due to other commitments I have had limited time to test this in all scenarios (both within Identity and use of the URF more generally).  I look forward to your feedback.

**Comment by saYRam** [comment]
Thank you for your labor. Now its works fine.
I'll try all scenarios as best i can and report back to you if something goes wrong.

**Comment by Neilski** [comment]
Thank you saYRam, if you find any problems, shout!  It was a shame to have to resort to editing the URF library, but I couldn't see another way around the problem without a lot of work and that was not really the purpose of the project.  The Identity related stuff seems OK, but pay particular attention to your normal URF activities to make sure that the changed logic in the SyncObjectsStatePreCommit() doesn't break any of the expected functionality.  Good luck!

**Comment by saYRam** [comment]
Hi Neilski. What ForceInitialLocalAccount for?

**Comment by Neilski** [comment]
In my particular application (where this project was derived from), the customer *always* wanted the user to create a local account (email address and password) before they could link that to an external oAuth provider.

The ForceInitialLocalAccount setting enables you to override this behaviour - although you will probably need to edit some of the logic in the AccountController to make the login process behave how you want it to.

**Comment by saYRam** [comment]
I set it true and i saw what it is. Thank you. When user uses external login a password not created.
I'm planning to let to user create password via profile page who registered with external login providers. With that way user can login with external provider or local account. Can using both external and local account be objectionable?

I'm not using email address as username btw.

**Comment by Neilski** [comment]
Yes, that should be fine - just adjust the AccountController logic and its associated views to give you the login/registration process that you require.  Out-of-the box, the project insists/expects a concrete local user before anything else, but in reality it is just Microsoft's Identity 2 example template with a few modifications, so you are free to change it however you want or indeed revert back to the default Microsoft Template implementation if that suits you better.
